plugins/mac-capture/mac-window-capture.m

```diff
@@ -13,71 +13,91 @@
 
 	struct cocoa_window window;
 
-	//CGRect              bounds;
-	//CGWindowListOption  window_option;
 	CGWindowImageOption image_option;
 
-	CGColorSpaceRef color_space;
-
 	DARRAY(uint8_t) buffer;
 
+	uint64_t last_tick;
+	struct obs_source_frame frame;
+
 	pthread_t capture_thread;
 	os_event_t *capture_event;
 	os_event_t *stop_event;
+
+	pthread_mutex_t mutex;
 };
 
 static CGImageRef get_image(struct window_capture *wc)
 {
-	NSArray *arr = (NSArray *)CGWindowListCreate(
-		kCGWindowListOptionIncludingWindow, wc->window.window_id);
-	[arr autorelease];
+	CFMutableArrayRef arr =
+		CFArrayCreateMutable(kCFAllocatorDefault, 0, NULL);
+	CFArrayAppendValue(arr, (void *)(uintptr_t)wc->window.window_id);
+
+	CGImageRef image = CGWindowListCreateImageFromArray(
+		CGRectNull, (CFArrayRef)arr, wc->image_option);
 
-	if (!arr.count && !find_window(&wc->window, NULL, false))
-		return NULL;
+	if (!image && find_window(&wc->window, NULL))
+		image = get_image(wc);
```

Comment:
I debated whether this recursive call was acceptable, but I personally would rather use a label and `goto` here.

The idea is if `CGWindowListCreateImageFromArray()` returns `NULL`, then `!image` would be true, which means that the window ID has changed, or at least the window ID we just tried to use failed to produce a `CGImage` for us, so let's try to find it by name again, thus the call to `find_window()` and if that returns `TRUE`, then let's try using this new `window_id` by calling `get_image()` again, recursively.

This isn't great, but it's good enough, and minimizes the calls to `find_window()` which is desirable.

Reply:
Are there "common" scenarios where this would happen? If the window ID has become invalid, then the capture configuration has become invalid as well, just matching it again by name and hoping for the best doesn't sound that great to me.

IMO the user should then re-configure the source to explicitly chose the window they want to capture as one reason to use window capture is to be specific(!) about which content is broadcast/recorded and not have OBS just pick another window without user input.

Reply:
The common scenario that I'm personally familiar with is using Window Capture for streaming League of Legends, where the game utilizes two windows, one for the main game client which is always present, and a window that is created when you enter into an actual game.  Window Capture being able to "watch for" that game window, and using Scene Switcher and other stuff to automatically switch between broadcasting the game window when you're in-game vs. the main client when you're not, is really convenient.

Reply:
https://github.com/obsproject/obs-studio/blob/master/plugins/win-capture/window-capture.c#L456

Generally windows has 3 capture priorities, by title, class or exe. Someone with more familiarity would have to make sure that it actually recaptures after the window is lost (such as closing it). If so, this seems ok for the default to bring it in line with windows.